[TwigBundle] Added priority to twig extensions

Brunty opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | Don't merge before the docs

Added priority to twig extensions in the `TwigEnvironmentPass` to control the order in which they're registered, similar to the `TwigLoaderPass`

Though, unsure on what priority to use as a default, and will PR docs when finalised.

Brunty committed:
[TwigBundle] Added priority to twig extensions

iltar commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/TwigEnvironmentPass.php:
```diff
@@ -28,16 +28,26 @@ public function process(ContainerBuilder $container)
+        foreach ($container->findTaggedServiceIds('twig.extension', true) as $id => $attributes) {
+            $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : 0;
+            $prioritizedLoaders[$priority][] = $id;
+        }
```
Could this be replaced by using the [`PriorityTaggedServiceTrait`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php)?

Brunty replied:
@iltar done :)

Brunty committed:
[TwigBundle] Refactor to PriorityTaggedServiceTrait

fabpot commented:
Can you explain the use case for having priorities on extensions? I do understand the need for loaders, but I fail to see why we would need them for extensions.

Brunty commented:
@fabpot I was recently working on something and needed to overload a filter - I was lucky in that where I was registering it, it was after the previous one being registered (so as per docs here: https://twig.symfony.com/doc/2.x/advanced.html#overloading) but figured that might not always be the case.

nicolas-grekas commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/TwigEnvironmentPass.php:
```diff
@@ -36,8 +39,8 @@ public function process(ContainerBuilder $container)
-        foreach ($container->findTaggedServiceIds('twig.extension', true) as $id => $attributes) {
-            $definition->addMethodCall('addExtension', array(new Reference($id)));
+        foreach ($this->findAndSortTaggedServices('twig.extension', $container) as $extension) {
+            $definition->addMethodCall('addExtension', array(new Reference($extension)));
```
$extension is already a reference so that this code could be simplified

Brunty committed:
[TwigBundle] Refactor extension variable as already a reference

Brunty committed:
[TwigBundle] Remove unused use statement

ro0NL reviewed (approved):
needs a CHANGELOG entry i guess.

ro0NL commented on src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Compiler/TwigEnvironmentPassTest.php:
```diff
@@ -0,0 +1,72 @@
+class TwigEnvironmentPassTest extends TestCase
+{
+    /**
+     * @var \PHPUnit_Framework_MockObject_MockObject
```
for clarity i prefer `/** @var Type */` thus inline. But not a blocker :)

nicolas-grekas replied:
AFAIK, we don't do much "inline" on properties.

ro0NL commented:
4.0.0 but there's a 4.4.0 entry already, i think it's a typo :)

Brunty commented:
@ro0NL I'll add a 4.0.0 section, want me to correct the 4.4.0 one to 4.0.0?

ro0NL commented:
think so. affects master only anyway.

Brunty committed:
[TwigBundle] Add changelog entry for priority on extensions

Brunty commented:
@ro0NL - added changelog entry for you

nicolas-grekas commented on src/Symfony/Bundle/TwigBundle/CHANGELOG.md:
```diff
@@ -1,10 +1,11 @@
 -----
 
  * removed `ContainerAwareRuntimeLoader`
+ * added priority to twig extensions 
```
should be added in a new entry for 4.1.0

Brunty replied:
@nicolas-grekas done :)

Brunty committed:
[TwigBundle] Move changelog entry to 4.1.0

xabbuh commented on src/Symfony/Bundle/TwigBundle/CHANGELOG.md:
```diff
@@ -1,7 +1,11 @@
-4.4.0
+4.1.0
+-----
+ * added priority to twig extensions 
```
Twig

Brunty replied:
Changed

xabbuh commented on src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Compiler/TwigEnvironmentPassTest.php:
```diff
@@ -0,0 +1,72 @@
+
+    protected function setUp()
+    {
+        $this->builder = $this->getMockBuilder('Symfony\Component\DependencyInjection\ContainerBuilder')->setMethods(array('hasDefinition', 'findTaggedServiceIds', 'setAlias', 'getDefinition'))->getMock();
```
I would not mock the `ContainerBuilder` class

Brunty replied:
I based it on this test: https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Compiler/TwigLoaderPassTest.php#L36

I can change it to use the actual builder if you'd like?

xabbuh replied:
please do so :)

Brunty replied:
@xabbuh done :)

Brunty committed:
[TwigBundle] Fix capitalisation in changelog

Brunty committed:
[TwigBundle] Remove mock container builder in test

Brunty committed:
[TwigBundle] Fix code style errors

nicolas-grekas commented on src/Symfony/Bundle/TwigBundle/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+4.1.0
+-----
+ * added priority to Twig extensions 
```
missing blank line before

Brunty replied:
Added :)

nicolas-grekas commented on src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Compiler/TwigEnvironmentPassTest.php:
```diff
@@ -0,0 +1,62 @@
+     */
+    private $pass;
+
+    protected function setUp()
```
we don't need this method IMHO, better remove it and the corresponding properties

Brunty replied:
Done :)

Brunty committed:
[TwigBundle] Update changelog for newline formatting

Brunty committed:
[TwigBundle] Remove TwigEngironmentPassTest setUp method

Moves the properties on the test class to variables inside the single test method it has.

Brunty committed:
[TwigBundle] Fix code style in TwigEnvironmentPassTest

fabpot commented:
Thank you @Brunty.
